File: Java/src/RemoveNthNodeFromEndofList.java

public class RemoveNthNodeFromEndofList {
	
	public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode node = head;
        int counter = 0;
        while(node != null) {
        	node = node.next;
        	counter++;
        }
        
        int index = counter - n;
        if(index == 0) return head.next;
        
        node = head;
        for(int i = 0; i < index - 1; i++) {
        	node = node.next;
        }
        
        node.next = node.next.next;
        
        return head;
        
    }
	
	public static void main(String[] args) {
		ListNode n1 = new ListNode(1);
		ListNode n2 = new ListNode(2);
		ListNode n3 = new ListNode(3);
		ListNode n4 = new ListNode(4);
		ListNode n5 = new ListNode(5);
		n1.next = n2;
		n2.next = n3;
		n3.next = n4;
		n4.next = n5;
		
		RemoveNthNodeFromEndofList test = new RemoveNthNodeFromEndofList();
		ListNode head = test.removeNthFromEnd(n1, 2);
		
		System.out.println("hello");
		
	}

}

File: Python/RemoveNthNodeFromEndofList.py
from ListNode import ListNode

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        count = 0
        cache = []
        node = head
        while(node != None):
            count = count + 1;
            cache.append(node)
            node = node.next;

        if count == n:
            return head.next
        index = count - n
        cache[index-1].next = cache[index-1].next.next
        return head


if __name__ == "__main__":
    test = Solution()
    n1 = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(3)
    n4 = ListNode(4)
    n5 = ListNode(5)
    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n5

    test.removeNthFromEnd(n1, 2)

    node = n1
    while(node != None):
        print(node.val)
        node = node.next

